sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c

```diff
@@ -478,77 +512,399 @@ static struct codec_info codec_info_list[] = {
+	int i;
+
+	for (i = 0; i < link->num_adr; i++) {
+		/* skip itself */
+		if (i == id)
+			continue;
+
+		sdw1_version = SDW_VERSION(link->adr[i]);
+		mfg1_id = SDW_MFG_ID(link->adr[i]);
+		part1_id = SDW_PART_ID(link->adr[i]);
+		class1_id = SDW_CLASS_ID(link->adr[i]);
+
+		if (sdw_version == sdw1_version &&
+		    mfg_id == mfg1_id &&
+		    part_id == part1_id &&
+		    class_id == class1_id)
+			return false;
+	}
+
+	return true;
+}
+
+static int create_codec_dai_name(struct device *dev,
+				 const struct snd_soc_acpi_link_adr *link,
+				 struct snd_soc_dai_link_component *codec)
+{
+	unsigned int sdw_version, unique_id, mfg_id;
+	unsigned int link_id, part_id, class_id;
+	int i, id;
+
+	for (i = 0; i < link->num_adr; i++) {
+		sdw_version = SDW_VERSION(link->adr[i]);
+		link_id = SDW_DISCO_LINK_ID(link->adr[i]);
+		unique_id = SDW_UNIQUE_ID(link->adr[i]);
+		mfg_id = SDW_MFG_ID(link->adr[i]);
+		part_id = SDW_PART_ID(link->adr[i]);
+		class_id = SDW_CLASS_ID(link->adr[i]);
+
+		if (is_unique_device(link, sdw_version, mfg_id, part_id,
+				     class_id, i))
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id);
+		else
+			codec[i].name = devm_kasprintf(dev, GFP_KERNEL,
+						       "sdw:%x:%x:%x:%x:%x",
+						       link_id, mfg_id, part_id,
+						       class_id, unique_id);
+
+		if (!codec[i].name)
+			return -ENOMEM;
+
+		id = find_codec_info(part_id, NULL);
+		if (id < 0)
+			return id;
+
+		codec[i].dai_name = codec_info_list[id].dai_name;
+	}
+
+	return 0;
+}
+
+static void set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
+				struct snd_soc_dai_link *dai_links,
+				bool playback)
+{
+	unsigned int part_id;
+	int i, id;
+
+	for (i = 0; i < link->num_adr; i++) {
+		part_id = SDW_PART_ID(link->adr[i]);
+		id = find_codec_info(part_id, NULL);
+
+		if (codec_info_list[id].init)
+			codec_info_list[id].init(link, dai_links,
+						 &codec_info_list[id],
+						 playback);
+	}
+}
+
+static int create_sdw_codec_dai(struct device *dev,
+				struct snd_soc_dai_link *dai_links, int id,
+				struct snd_soc_dai_link_component *cpus,
+				const struct snd_soc_acpi_link_adr *link)
+{
+	struct snd_soc_dai_link_component *codec;
+	unsigned int part_id, link_id;
+	int i, j = 0, idx;
+	int ret;
+
+	codec = devm_kcalloc(dev, link->num_adr,
+			     sizeof(struct snd_soc_dai_link_component),
+			     GFP_KERNEL);
+	if (!codec)
+		return -ENOMEM;
+
+	ret = create_codec_dai_name(dev, link, codec);
+	if (ret < 0)
+		return ret;
+
+	part_id = SDW_PART_ID(link->adr[0]);
+	idx = find_codec_info(part_id, NULL);
+	if (idx < 0)
+		return idx;
+
+	link_id = ffs(link->mask) - 1;
+
+	/* playback & capture */
+	for (i = 0; i < 2; i++) {
+		char *name, *cpu_name;
+
+		if (!codec_info_list[idx].direction[i])
+			continue;
+
+		name = devm_kasprintf(dev, GFP_KERNEL,
+				      sdw_stream_name[i], link_id);
```

Comment:
isn't this function called only once in the SOF life-cycle? If yes - you don't need `sdw_stream_name[]` as a static array, you can put it into this function

Reply:
done, thanks!